[#5152] Write preprocessed P4 to `repro.p4` file when `-P` option is provided

kfcripps opened the pull request:
Closes #5152.

Marking as draft until we decide:

- What to name the option - I have named it `-P` (for "Preprocessed") for now, but @asl also suggested `--save-temps`.
- What to name the generated file - I named it `repro.p4` (short for "reproducer"), but maybe there is a better name.

asl commented:
Not sure why the output should be named `repro.p4`. Normally for gcc / clang we're having:
 - Option is named `--save-temps`
 - For an input file named "foo.c" the preprocessed output is named `foo.i`.

kfcripps commented:
> Not sure why the output should be named `repro.p4`. Normally for gcc / clang we're having:
> 
> * Option is named `--save-temps`
> * For an input file named "foo.c" the preprocessed output is named `foo.i`.

Sure, that sounds reasonable to me.

ChrisDodd commented:
Isn't there already a `-E` option that just runs the preprocessor and writes that to stdout?

kfcripps commented:
@ChrisDodd Yes, but the compiler exits immediately after doing so - the purpose of this option is to be able to save the preprocessed P4 and continue with compilation instead of exiting.

vgurevich commented:
Not sure if it matters, but the existing Tofino implementation names this file as `foo.p4pp`

vgurevich commented:
The other thing I'd strongly recommend if you want to do it in the frontend: sanitize the preprocessed file by **not** including the standard system includes (i.e. `core.p4` and/or your own architectural files), or at least have an option to do that. This will allow to reproduce old issues much easier or notice compatibility issues whenever you change your arch files. Believe me, you'd like you did it.

kfcripps commented:
> Not sure if it matters, but the existing Tofino implementation names this file as `foo.p4pp`

Giving the generated preprocessed file a `.p4pp` suffix sounds good to me.

> The other thing I'd strongly recommend if you want to do it in the frontend: sanitize the preprocessed file by not including the standard system includes (i.e. core.p4 and/or your own architectural files), or at least have an option to do that. That means the file will have the user's code preprocessed, but will still use #include <core.p4> and #include <arch.p4>.

> This will allow to reproduce old issues much easier or notice compatibility issues whenever you change your arch files. Believe me, you'd like you did it.

I currently have no strong motivation to do that, but I am not opposed if someone else wants to add an option to do that in the future.

vlstill commented:
My opinion is that `--save-temps` sounds like a reasonable option and I would say `.p4pp` is better than just `.p4`.

I have no strong opinion about leaving system includes, although having it configurable (maybe with default being "leaving" `#include`) could be best. The disadvantage of desugaring back to `#include` is that it can hide modifications (which is probably irrelevant for just preprocessing) and obviously it excludes the text of the includes so some issues might be accidentally resolved by later changes.

vgurevich commented on frontends/common/parser_options.cpp:
```diff
@@ -455,19 +476,25 @@ std::optional<ParserOptions::PreprocessorResult> ParserOptions::preprocess() con
-    newName += baseSuffix;
-    newName += name.extension();
+    if (savePreprocessed) {
+        std::filesystem::path fileName = makeFileName(dumpFolder, "repro.p4", "");
```
I think that using `<program_name>.p4pp` would be better, compared to having a fairly arbitrary fixed name.

vgurevich commented on frontends/common/parser_options.h:
```diff
@@ -81,6 +81,8 @@ class ParserOptions : public Util::Options {
     /// if true preprocess only
     bool doNotCompile = false;
+    /// if true save preprocessed P4 to repro.p4
+    bool savePreprocessed = false;
```
If you decide to address my first comment, then, obviously, this comment needs to be changed too.

kfcripps commented:
Updates:

- Renamed the option to `--save-temps`
- Renamed the generated file to `filename.p4pp`
- Original code was broken because `in` was consumed for `repro.p4` generation, and was empty by the time actual program parsing occurred. The new code runs `cpp` twice and uses a second `FILE` stream for `filename.p4pp` generation instead.

vlstill reviewed (commented):
I think the double preprocessing is a bit unfortunate (it is wasteful and in an extreme case, they could produce different results because of changes on the filesystem). Maybe instead if we have `--save-temps` we should direct the preprocessor output to the `p4pp` file and then open that file for the parser.

kfcripps commented on frontends/common/parser_options.cpp:
```diff
@@ -421,16 +430,30 @@ const char *ParserOptions::getIncludePath() const {
 #else
-        std::string cmd("cpp");
+        cmd = "cpp";
 #endif
```
I moved the `cmd` declaration back above, but I'm not sure that macros can be used inside of regular `if` statements / ternary operations. When I tried doing this, I got the error:
```
../../../frontends/common/parser_options.cpp: In member function 'std::optional<std::unique_ptr<_IO_FILE, void (*)(_IO_FILE*)> > P4::ParserOptions::preprocess() const':
../../../frontends/common/parser_options.cpp:435:27: error: '__clang__' was not declared in this scope
  435 |         std::string cmd = __clang__ ? "cc -E -x c -Wno-comment" : "cpp";
      |                           ^~~~~~~~~
ninja: build stopped: subcommand failed.
```

kfcripps commented:
> I think the double preprocessing is a bit unfortunate (it is wasteful and in an extreme case, they could produce different results because of changes on the filesystem). Maybe instead if we have `--save-temps` we should direct the preprocessor output to the `p4pp` file and then open that file for the parser.

This does seem better to me too. Done.

kfcripps commented on frontends/common/parser_options.cpp:
```diff
@@ -405,16 +414,28 @@ const char *ParserOptions::getIncludePath() const {
     } else {
 #ifdef __clang__
-        std::string cmd("cc -E -x c -Wno-comment");
+        std::string cmd = "cc -E -x c -Wno-comment";
```
```suggestion
        std::string cmd("cc -E -x c -Wno-comment");
```

kfcripps commented on frontends/common/parser_options.cpp:
```diff
@@ -405,16 +414,28 @@ const char *ParserOptions::getIncludePath() const {
+        std::string cmd = "cc -E -x c -Wno-comment";
 #else
-        std::string cmd("cpp");
+        std::string cmd = "cpp";
```
```suggestion
        std::string cmd("cpp");
```
